src/Response.php

```diff
@@ -180,18 +181,14 @@ public static function fromString($string)
 
         $response = new static();
 
-        $regex   = '/^HTTP\/(?P<version>1\.[01]) (?P<status>\d{3})(?:[ ]+(?P<reason>.*))?$/';
-        $matches = [];
-        if (! preg_match($regex, $firstLine, $matches)) {
-            throw new Exception\InvalidArgumentException(
-                'A valid response status line was not found in the provided string'
-            );
+        try {
+            $response->parseStatusLine($firstLine);
+        } catch (RuntimeException $e) {
+            array_shift($lines); // skip next empty line
+            $firstLine = array_shift($lines); // and try again
+            $response->parseStatusLine($firstLine);
```

Comment:
Make sure next line is empty before skipping.
On a quick glance at the specification I do not see anything preventing sending more headers with 100 status code before proceeding with actual response.

Also your extracted method does too much. Extract status line parsing logic but leave the rest out of it.
I would do something like this:

```php
list($version, $status, $reason) = $this->parseStatusLine($firstLine);
if (static::STATUS_CODE_100 === $status) {
    $next = array_shift($lines);
    if (!empty($next)) {
        // throw exception, probably
    }
    $next = array_shift($lines);
    list($version, $status, $reason) = $this->parseStatusLine($next);
}

$response->version = $version;
$response->setStatusCode($status);
$response->setReasonPhrase($reason);
```

Reply:
Changed to work without exceptions. Rebased and fixed tests.